### PR TITLE
Release automation script

### PR DESCRIPTION
**Pull request by kislyuk**
Automates releases of code from master to staging and from staging to prod (and between any two other branches, as long as an environment config file exists).

Checks for CI status and allows bypass.
Checks for commits only on the target branch.

I'm particularly interested in your review of the git commands. They include a delightful `git push --force` :shipit: 

Connects to #467

### CODE REVIEW

**Comment by ttung** [comment]
Generally speaking, I think we should be able to know which version of the code we were running at any given point.  Force-pushing destroys that.

Here's my suggestion:

1. Have a `cut` operation, which moves a `prod`/`staging` branch to a fixed location.
2. Revisions that need to be cherry-picked onto the `prod`/`staging` branch can be done accordingly.
3. Each time we deploy, we tag back on git.

Thoughts?

**Comment by kislyuk** [comment]
I agree we should know which code we were running at any given point. The deployment tags do accomplish that: commits referred to by the tags will stay in the tree, even if their underlying branch gets deleted or force pushed.

I'm not sure if I understand your suggestion. How does the `cut` operation sidestep the need for force-pushing?

Cherry-picking and custom bugfixes applied to prod and staging can routinely cause it not to be cleanly fast-forwardable to the head of master. In those situations I think I prefer blowing away the release branch and branching it from master again, instead of dealing with history differences. Since we tag every release, blowing away the named branch doesn't destroy information.

**Comment by ttung** [comment]
I completely and utterly misread your diff. I do not know what happened. :(

Anyway, I think what you have makes a lot of sense. Please ignore my top-level comment, and just pay attention to the line-level comments. :)

**Comment by kislyuk** [comment]
Hi @ttung, I don't see line-level comments, can you check that they're not still pending with the review or anything like that? Thanks!

**Comment by ttung** [comment]
Whoops. :(

File: scripts/release.sh [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    else
+        echo "Status checks failed on branch $PROMOTE_FROM_BRANCH."
+        echo "Run with --force to promote $PROMOTE_FROM_BRANCH to $PROMOTE_DEST_BRANCH and deploy anyway."
+        exit 1
+    fi
+fi
+
+RELEASE_TAG=${PROMOTE_DEST_BRANCH}-$(date -u +"%Y-%m-%d-%H-%M-%S").deploy
+
+if [[ "$(git --no-pager history $PROMOTE_DEST_BRANCH ^$PROMOTE_FROM_BRANCH --no-merges)" != "" ]]; then
```
[the remark this answers is not in the document]
Comment (kislyuk) [reply scripts/release.sh]: Yes! Thanks, fixed.

File: scripts/release.sh [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        echo -e "\nThey will be overwritten on $PROMOTE_DEST_BRANCH and discarded."
+    else
+        echo -e "\nRun with --force to overwrite and discard these commits from $PROMOTE_DEST_BRANCH."
+        exit 1
+    fi
+fi
+
+git fetch --all
+git checkout origin/$PROMOTE_FROM_BRANCH
+git checkout -B $PROMOTE_DEST_BRANCH
```
[the remark this answers is not in the document]
Comment (kislyuk) [reply scripts/release.sh]: Added a clause that checks for tracked modified content and stops the release if found.

**Pull request merged**